create storage migration adr (#4196)
replicatedhq/kURL: docs/arch/adr-009-storage-migration.md

## Context

Single node installs should not use Rook, as doing so adds complexity, increases resource usage, and invites disk corruption.
Multi node installs very often require distributed storage, best provided by Rook.

Many customers will move from a single-node install to a multi-node install as circumstances change.
This migration should be as simple as possible.
Vendors generally should also not need to specify multiple installer specs for single and multi-node installs.

## Decision

Provide a migration path from OpenEBS to OpenEBS + Rook, triggered by the presence of 3+ nodes.

## Solution

A new field, "minimum node count", will be added to the rook config. (minimum value 2, requires openebs localpv to be present)
When it is set, the rook operator will be installed on single node installs, but no ceph cluster will be created.

An example spec is provided below:

```yaml
spec:
  rook:
    version: "1.10.x"
    storageClassName: "distributed"
    minimumNodeCount: 3
  openebs:
    version: "3.4.x"
    isLocalPVEnabled: true
    localPVStorageClassName: "local"
```

When installed on a single node, two storageclasses will be available - "scaling" and "local". 

> The default storage class will be "scaling", and both storage classes will be backed by openebs localpv.

Reviewer: How is this going to work in currently installed cluster where the user has named OpenEBS storage class as `foo` and wants to have its data automatically migrated to `distributed` ?

Author: it'll be based on whatever openebs storageclass name you have specified. The only one that isn't configurable is `scaling`, and that will be created upon updating to this kurl version with the appropriate config.

If someone for some reason has an existing storageclass named `scaling`, I'm happy to say I'll burn that bridge when I get to it.

> When the third node is added, a ceph cluster will be created by the EKCO operator.

Reviewer: So Rook won't be deployed by the `installer` whenever `minimumNodeCount` is specified ?

Reviewer: Note: installer still needs to load the images in all nodes.

Author: Rook will be deployed, but just the operator - there won't be a cephcluster or storageclass to begin with (and thus no OSD pods etc)

When that ceph cluster becomes healthy (with at least 3 replicas), the "distributed" storageclass will be created using rook-ceph.
pvmigrate will then be used to migrate all data from the "scaling" storageclass to the "distributed" storageclass, and the default storageclass will be changed to "distributed".

> This process does involve stopping pods using "scaling" storage.

Reviewer 2: ```suggestion
This process does involve downtime, stopping pods using "scaling" storage.
```

In this way, applications can specifically request storage that will always be local to a node (with the "local" storageclass), or storage that will be distributed across the cluster (with the "distributed" storageclass).
Using the "scaling" storageclass directly (instead of merely using the default storageclass) would be an application linting error.

YET TO BE DECIDED:
1. Should the migration be triggered immediately upon adding a third node, or should it be triggered by the user? It can involve application downtime!

>    1. Allow the user to trigger the migration with a 'tasks.sh migrate-storage' command.

Reviewer: I prefer this approach.

Author: that's what I'm going with (along with kotsadm trigger)

>    2. Provide a prompt in kotsadm that will trigger the migration.

Reviewer 2: should you mention something about warning the user with some sort of banner in the kots ui that the cluster is running in an unstable state prior to a storage migration?

Author: yeah, will do

   3. Trigger the migration automatically when the third node is added.

> 2. What to do with object storage. (should data be migrated from MinIO to Rook, and MinIO deleted?)

Reviewer 2: I believe the solution must also take into account vendors who do not require distributed storage such as swimlane. Can you go into a little more detail in the soution about how the cluster is configured in this use case? For example, is minio required to back the registry and velero?

Author: I would say that this proposal is almost completely irrelevant to vendors that don't require distributed storage!

That being said, I've added automatic minio migration to this doc.

## Decision

Provide a migration path from OpenEBS to OpenEBS + Rook, triggered by the presence of 3+ nodes.

## Solution

A new field, "minimum node count", will be added to the rook config. (minimum value 2, requires openebs localpv to be present)
When it is set, the rook operator will be installed on single node installs, but no ceph cluster will be created.

An example spec is provided below:

```yaml
spec:
  rook:
    version: "1.10.x"
    storageClassName: "distributed"
    minimumNodeCount: 3
  openebs:
    version: "3.4.x"
    isLocalPVEnabled: true
    localPVStorageClassName: "local"
```

When installed on a single node, two storageclasses will be available - "scaling" and "local". 
The default storage class will be "scaling", and both storage classes will be backed by openebs localpv.

When the third node is added, a Ceph cluster will be created by the EKCO operator.
When that Ceph cluster becomes healthy (with at least 3 replicas), the "distributed" storageclass will be created using rook-ceph.
However, a migration will not begin until one of three things occurs:

> 1. The user approves the migration in kotsadm

Reviewer 2: ```suggestion
1. The user approves the migration in KOTS
```

> 2. The user runs `install.sh` on a primary node, and accepts a prompt to migrate storage

Reviewer 2: Can we do any trickery to prompt and migrate the storage on a secondary node? This seems particularly useful for the use case where the user inits a cluster and immediately joins two additional worker nodes before installing a KOTS app.

> 2. The user runs `install.sh` on a primary node, and accepts a prompt to migrate storage

Author: For both Ethan and Kyle - I would love a way to trigger this upon adding the third node, even if it's not a primary node.

I just don't see a way to do this besides either:
1. adding a http endpoint, and having the script curl that after the node has been joined
2. having a daemonset watching files in a directory on the host

and both of those seem a little irritating, but in retrospect the only real irritation from the http endpoint would be authentication and choosing an appropriate port. (And we could include a token in the join script + use a dynamic port, so that's doable)

> 2. The user runs `install.sh` on a primary node, and accepts a prompt to migrate storage

Reviewer 2: i think we need to answer this question. i cannot get onboard with this proposal without some way to bootstrap a 3 node cluster and not have to go through a storage migration.

> 2. The user runs `install.sh` on a primary node, and accepts a prompt to migrate storage

Author: I think an easy solution would be to do an automatic migration if no app has been installed yet, or if the app doesn't use any PVCs in the scaling storageclass - that way we only require customer approval for things that will take the app down. (things that will take kotsadm down are different, IMO)

> 3. The user runs the `migrate-multinode-storage` command in `tasks.sh` from a primary node

Reviewer 2: I'd prefer that this be a go command directly so that it is reusable. If the only interface we have for the end user is tasks.sh then the shell should only invoke a go command and nothing else.

Author: Agreed that the logic in tasks.sh should be minimal-to-nonexistent

The migration process will NOT be triggered automatically.

> While the migration process is available to be triggered, a banner will be shown in Kotsadm indicating that the migration is available, and that the cluster is currently in an undesirable state.

Reviewer 2: ```suggestion
While the migration process is available to be triggered, a banner will be shown in the KOTS browser UI indicating that the migration is available, and that the cluster is currently in an undesirable state.
```

The migration process will be as follows:

> A trigger configmap will be created in the `kurl` namespace, to be observed by EKCO, prompting it to carry out the following steps.

Reviewer 2: Can we make use of a CRD? This may be new to the project but it is about time.

Author: it might actually be better to make it a http endpoint, so that join.sh could hit it too

Reviewer 2: when you come to a decision will you change this?

Author: Yeah, I'll change this to the HTTP endpoint option

> If MinIO is present, Kotsadm will be scaled down, and the existing `sync-object-store` command will be used to migrate data from MinIO to Rook.

Reviewer 2: ```suggestion
If MinIO is present, KOTS will be scaled down, and the existing `sync-object-store` Go command will be used to migrate data from MinIO to Rook.
```

> Kotsadm, Registry, and Velero will then be updated to use the Rook object store, and Kotsadm scaled back up.

Reviewer 2: ```suggestion
KOTS, Registry, and Velero will then be updated to use the Rook object store, and KOTS scaled back up.
```

After MinIO data is migrated and its consumers updated, the MinIO statefulset and namespace will be deleted.
`pvmigrate` will then be used to migrate all data from the "scaling" storageclass to the "distributed" storageclass, and the default storageclass will be changed to "distributed".
This process does involve downtime, caused by stopping pods using "scaling" storage, and is why we require the migration to be manually initiated.

> In this way, applications can specifically request storage that will always be local to a node (with the "local" storageclass), or storage that will be distributed across the cluster (with the "distributed" storageclass).

Author: The idea is that 'distributed' will always be on Rook, and 'scaling' will always be on openebs-to-be-moved-to-rook. 

If we rename back to 'scaling', there's no way for a vendor to say "I don't want this to run until it's distributed storage".

Using the "scaling" storageclass directly (instead of merely using the default storageclass) would be an application linting error.

## Status

Proposed

## Consequences

Vendors will be able to specify a single spec for single and multi-node installs, with a smooth migration path between the two.

There will be short application downtime migrating between the two storage systems.
This could be mitigated by changing pvmigrate to not stop all pods at once, but it is as yet unknown how this would be accomplished.

> Vendor applications will not be able to use the kurl-provided object store with this configuration, as this object store will change endpoints during the migration process.

Reviewer 2: ```suggestion
Vendor applications will not be able to use the kurl-provided object store with this configuration, as this object store will change endpoints during the migration process and not preserve user credentials.
```

> It is believed this will not impact any vendors.

Author: Theoretically they could use rook-ceph (when it becomes available)

But if they depend on it, then why go through the trouble of migrating from openebs when they could just require Rook from the beginning?

# ADR 9: OpenEBS to OpenEBS + Rook Migration

## Context

Single node installs should not use Rook, as doing so adds complexity, increases resource usage, and invites disk corruption.
Multi node installs very often require distributed storage, best provided by Rook.

Many customers will move from a single-node install to a multi-node install as circumstances change.
This migration should be as simple as possible.
Vendors generally should also not need to specify multiple installer specs for single and multi-node installs.

## Decision

Provide a migration path from OpenEBS to OpenEBS + Rook, triggered by the presence of 3+ nodes.

## Solution

A new field, "minimum node count", will be added to the rook config. (minimum value 2, requires openebs localpv to be present)
When it is set, the rook operator will be installed on single node installs, but no ceph cluster will be created.

An example spec is provided below:

```yaml
spec:
  rook:
    version: "1.10.x"
    storageClassName: "distributed"
    minimumNodeCount: 3
  openebs:
    version: "3.4.x"
    isLocalPVEnabled: true
    localPVStorageClassName: "local"
```

> When installed on a single node, two storageclasses will be available - "scaling" and "local". 

Reviewer 3: Calling it _scaling_ doesn't seem to convey it's purpose clear enough for me. Perhaps we can call it _centralized_ which is the opposite of _distributed_? _Centralized_ makes sense to me as it convey cluster storage that is currently designated for this node aside from calling it something like 'local 2`/ 'second local`.

Reviewer 2: how about "dynamic"?

Reviewer 2: "elastic"?

Author: 'dynamic' would be fine with me

Author: I'm going to merge this now though, and we can make changes to it afterwards

Reviewer 3: `dynamic` is better but I still think looking at the SC list, after a third node is added, that it better conveys the state of the cluster (i.e. that we went from non-distributed to distributed storage).
```
kubectl get sc
NAME                        PROVISIONER                                                RECLAIMPOLICY   VOLUMEBINDINGMODE      ALLOWVOLUMEEXPANSION   AGE
local                       openebs.io/local                                           Delete          WaitForFirstConsumer   false                  16d
centralized                 openebs.io/local                                           Delete          WaitForFirstConsumer   false                  16d
distributed (default)       rook-ceph.rbd.csi.ceph.com                                          
```
